lzy/s3-sink/src/main/java/ai/lzy/kafka/Job.java

```diff
@@ -0,0 +1,293 @@
+
+
+    private final ByteBuffer s3ChunkBuffer = ByteBuffer.allocate(BUFFER_SIZE);
+
+    @Nullable private final Consumer<String, byte[]> consumer;
+    @Nullable private final S3AsyncClient storageClient;  // Support only s3 for now
+
+    @Nullable private String multipartId;
+    @Nullable private int partNumber = 1;
+
+
+    @Nullable private Iterator<ConsumerRecord<String, byte[]>> resultsStream;
+
+    @Nullable CompletableFuture<UploadPartResponse> uploadAwaitable;
+    private final String bucket;
+    private final String key;
+
+    public record PollResult(
+        boolean completed,
+        @Nullable Status error,  // Can be not set if not completed
+        Duration pollAfter
+    ) {}
+
+    enum State {
+        Created,
+        AwaitingData,
+        SchedulingUpload,
+        AwaitingUpload,
+        Completed
+    }
+
+    public Job(KafkaHelper helper, KafkaS3Sink.SubmitRequest request) {
+        this.request = request;
+
+        if (!request.getStorageConfig().hasS3()) {
+            LOG.error("{} Failed to upload data to s3: Supports only s3 storage for now", this);
+
+            throw Status.UNIMPLEMENTED.withDescription("Supports only s3 storage for now").asRuntimeException();
+        }
+
+        var accessToken = request.getStorageConfig().getS3().getAccessToken();
+        var secretToken = request.getStorageConfig().getS3().getSecretToken();
+        var endpoint = request.getStorageConfig().getS3().getEndpoint();
+
+        var props = helper.toProperties();
+        props.put("enable.auto.commit", "false");
+        props.put("group.id", UUID.randomUUID().toString());
+
+        consumer = new KafkaConsumer<>(props);
+
+        var hostAndPort = HostAndPort.fromString(endpoint);
+
+        try {
+            storageClient = S3AsyncClient.builder()
+                .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create(accessToken,
+                    secretToken)))
+                .region(Region.US_WEST_1)
+                .endpointOverride(new URI("http", null, hostAndPort.getHost(), hostAndPort.getPort(), null, null, null))
+                .forcePathStyle(true)
+                .build();
+        } catch (URISyntaxException e) {
+            LOG.error("{} Provided bad endpoint: {}: ", this, endpoint, e);
+            throw Status.INVALID_ARGUMENT.asRuntimeException();
+        }
+
+        var uri = new AmazonS3URI(request.getStorageConfig().getUri());
+
+        this.bucket = uri.getBucket();
+        this.key = uri.getKey();
+
+        final CreateMultipartUploadResponse resp;
+        try {
+            resp = storageClient.createMultipartUpload(CreateMultipartUploadRequest.builder()
+                    .bucket(bucket)
+                    .key(key)
+                    .contentType("text/plain")
+                    .build())
+                .get();
+        } catch (InterruptedException | ExecutionException e) {
```

Comment:
we should retry all 5xx errors

Reply:
S3Client will automaticly retry errors from s3